File: CamStudy/CamStudy/ViewModels/StatisticsViewModel.swift
import Foundation

struct DailyStudyTotal: Identifiable {
    var id: Date { date }
    var date: Date
    var seconds: Int
}

@MainActor
final class StatisticsViewModel: ObservableObject {
    @Published var sessions: [StudySession] = []
    @Published var errorMessage: String?
    @Published var isLoading = false

    private let firestoreService = FirestoreService.shared
    private let calendar = Calendar.current

    func loadSessions(userId: String) {
        isLoading = true
        errorMessage = nil
        firestoreService.fetchStudySessions(userId: userId) { [weak self] result in
            Task { @MainActor in
                guard let self else { return }
                self.isLoading = false
                switch result {
                case .success(let sessions):
                    self.sessions = sessions
                case .failure(let error):
                    self.errorMessage = error.localizedDescription
                }
            }
        }
    }

    var todaySeconds: Int {
        sessions
            .filter { calendar.isDateInToday($0.startedAt) }
            .reduce(0) { $0 + $1.durationSeconds }
    }

    var last7Days: [DailyStudyTotal] {
        let today = calendar.startOfDay(for: Date())
        return (0..<7).reversed().map { offset in
            let day = calendar.date(byAdding: .day, value: -offset, to: today) ?? today
            let seconds = sessions
                .filter { calendar.isDate($0.startedAt, inSameDayAs: day) }
                .reduce(0) { $0 + $1.durationSeconds }
            return DailyStudyTotal(date: day, seconds: seconds)
        }
    }
}

File: camstudy_app/lib/viewmodels/statistics_view_model.dart
import 'package:flutter/foundation.dart';

import '../models/study_session.dart';
import '../services/firestore_service.dart';

class DailyStudyTotal {
  final DateTime date;
  final int seconds;

  const DailyStudyTotal({required this.date, required this.seconds});
}

class StatisticsViewModel extends ChangeNotifier {
  List<StudySession> sessions = [];
  String? errorMessage;
  bool isLoading = false;

  final _firestoreService = FirestoreService.shared;

  Future<void> loadSessions(String userId) async {
    isLoading = true;
    errorMessage = null;
    notifyListeners();

    try {
      sessions = await _firestoreService.fetchStudySessions(userId);
    } catch (e) {
      errorMessage = e.toString();
    }
    isLoading = false;
    notifyListeners();
  }

  int get todaySeconds {
    final now = DateTime.now();
    return sessions
        .where((session) => _isSameDay(session.startedAt, now))
        .fold(0, (sum, session) => sum + session.durationSeconds);
  }

  List<DailyStudyTotal> get last7Days {
    final now = DateTime.now();
    final today = DateTime(now.year, now.month, now.day);
    return List.generate(7, (index) {
      final day = today.subtract(Duration(days: 6 - index));
      final seconds = sessions
          .where((session) => _isSameDay(session.startedAt, day))
          .fold(0, (sum, session) => sum + session.durationSeconds);
      return DailyStudyTotal(date: day, seconds: seconds);
    });
  }

  bool _isSameDay(DateTime a, DateTime b) {
    return a.year == b.year && a.month == b.month && a.day == b.day;
  }
}
